README
INTRODUCTION:

pequel -- Generate Perl programs for data transformation processes.

Pequel is a comprehensive data processing system for raw (ASCII) data file processing. 
It features a simple, user-friendly event driven scripting interface that transparently 
generates, builds and executes highly efficient data-processing programs. By using the 
Pequel scripting language, the user can create and maintain complex ETL data transformation 
processes quickly, easily, and accurately.  


Pequel is hosted on sourceforge: 

http://sourceforge.net/projects/pequel/



Installation Instructions
-------------------------

perl Makefile.PL

-or to specify different perl library path-

perl Makefile.PL PREFIX=/product/perldev/Perl/Modules

make

make test

make install


Documentation
-------------
The lastest documentation and package may be found at:

http://sourceforge.net/projects/pequel/



Installation troubleshooting
----------------------------

When installing into non-default directory, i.e., if you used the PREFIX, then you need to (probably) set the
following environment before 'make install'

PERL_INSTALL_ROOT -- set this to whatever you specified for PREFIX above.


...and the following before executing 'pequel'

PERL5LIB -- note the Installing messages displayed during the 'make install', and set this to the path up to and excluding
'Pequel'.

PATH -- add the directory containing the Pequel executable to the PATH variable. 
Note the installation messages for '.../bin/pequel'; add this path to the PATH environment variable.





Example installation
--------------------

> perl Makefile.PL PREFIX=/usr/local/Perl
Checking if your kit is complete...
Looks good
Writing Makefile for ETL::Pequel

> make
Skip lib/ETL/Pequel.pm (unchanged)
Skip lib/ETL/Pequel/Main.pm (unchanged)
Skip lib/ETL/Pequel/Param.pm (unchanged)
Skip lib/ETL/Pequel/Type.pm (unchanged)
Skip lib/ETL/Pequel/Script.pm (unchanged)
Skip lib/ETL/Pequel/Field.pm (unchanged)
Skip lib/ETL/Pequel/Error.pm (unchanged)
Skip lib/ETL/Pequel/Engine.pm (unchanged)
Skip lib/ETL/Pequel/Engine/Inline.pm (unchanged)
Skip lib/ETL/Pequel/Collection.pm (unchanged)
Skip lib/ETL/Pequel/Code.pm (unchanged)
Skip lib/ETL/Pequel/Docgen.pm (unchanged)
Skip lib/ETL/Pequel/Parse.pm (unchanged)
Skip lib/ETL/Pequel/Pod2Pdf.pm (unchanged)
Skip lib/ETL/Pequel/Lister.pm (unchanged)
Skip lib/ETL/Pequel/Table.pm (unchanged)
Skip lib/ETL/Pequel/Type/Date.pm (unchanged)
Skip lib/ETL/Pequel/Type/Section.pm (unchanged)
Skip lib/ETL/Pequel/Type/Option.pm (unchanged)
Skip lib/ETL/Pequel/Type/Macro.pm (unchanged)
Skip lib/ETL/Pequel/Type/Aggregate.pm (unchanged)
Skip lib/ETL/Pequel/Type/Db.pm (unchanged)
Skip lib/ETL/Pequel/Type/Db/Oracle.pm (unchanged)
Skip lib/ETL/Pequel/Type/Db/Sqlite.pm (unchanged)
Skip lib/ETL/Pequel/Type/Table.pm (unchanged)
Skip lib/ETL/Pequel/Type/Table/Oracle.pm (unchanged)
Skip lib/ETL/Pequel/Type/Table/Sqlite.pm (unchanged)
Manifying blib/man3/ETL::Pequel::Pod2Pdf.3

> export PERL_INSTALL_ROOT=/usr/local/Perl
> make test
t/01_aggregates_1........ok
t/02_array_fields........ok
t/03_conditional_aggr....ok
t/04_filter_regex........ok
t/05_group_by_derived....ok
t/06_group_by_derived_2..ok
t/07_hash_option.........ok
t/08_local_table.........ok
t/09_macro_select........ok
t/10_output_calc_fields..ok
t/11_statistics_aggr.....ok
t/12_statistics_aggr_2...ok
t/13_transfer_option.....ok
t/14_simple_tables.......ok
t/15_external_tables.....ok
t/16_sales_ttl_by_loc....ok
t/17_pequel_tables.......ok
t/18_chain_pequel........ok
t/19_divert_record.......ok
t/20_copy_record.........ok
t/21_copy_output.........ok
t/22_output_combiner.....ok
All tests successful.
Files=22, Tests=22, 155 wallclock secs (136.11 cusr + 19.73 csys = 155.84 CPU)

> make install
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Param.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Code.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Collection.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Docgen.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Engine.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Error.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Field.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Lister.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Main.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Parse.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Pod2Pdf.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Script.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Table.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Aggregate.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Date.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Db.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Macro.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Option.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Section.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Table.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Table/Oracle.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Table/Sqlite.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Db/Oracle.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Type/Db/Sqlite.pm
Installing /usr/local/Perl/usr/perl5/site_perl/5.6.1/ETL/Pequel/Engine/Inline.pm
Installing /usr/local/Perl/usr/local/Perl/bin/pequel
Installing /usr/local/Perl/usr/local/Perl/bin/pequelpod2pdf
Installing /usr/local/Perl/usr/local/Perl/man/man3/Pequel::Pod2Pdf.3
Writing /usr/perl5/site_perl/5.6.1/sun4-solaris-64int/auto/Pequel/.packlist
Appending installation info to /usr/local/Perl/lib/sun4-solaris-64int/perllocal.pod

> export PERL5LIB=/usr/local/Perl/usr/perl5/site_perl
> export PATH=$PATH:/usr/local/Perl/usr/local/Perl/bin
> pequel -v
pequel Version 2.3-6, Build: Monday October  24 23:16:49 BST 2005


Copyright Notice
----------------
Copyright 1999-2005 Mario Gaffiero.

This file is part of Pequel(TM).

Pequel is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

Pequel is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Pequel; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


Changes
Revision history

* Version 2.4-6b -- 3 March 2006

- Fixed tests on certain platforms that 1) don't support 'cat', and 2) have non-standard
	perl exec location.

* Version 2.4-6 -- 14 February 2006

- New Pequel Quick Reference Guide (docs/pequelquickref.pdf).
- Fixed module name in ETL::Pequel::Pod2Pdf, scripts/pequelpod2pdf.
- Option: exec_min_lines -- use this to specify max program code size that will execute with eval() 
	otherwise execute with exec() on external tmp EXEC program file. Defaults to 200 lines. The exec() 
	execution will substantially improve the performance when copy/divert sections are in use.
- Fixed user defined date type parse problem.
- Some vim-syntax updates.

* Version 2.4-5 -- 16 November 2005

- Fixed test failures (on some platforms) -- caused by extra pipe ('|') in open() in code generated for
	sort-output section.
- New option 'show_synonyms' -- replace field synonym by field-number in generated code.
	-- set to zero for additional processing performance.
- Bug fix numeric/decimal type checking for nulls/nonulls options.
- Bug fix numeric/decimal type comparison in dedup-on.
- Bug fix -- input record line counter.

* Version 2.4-4 -- 3 November 2005

- Updated User Guide to include perl module ETL::Pequel.pm interface reference.
- Bug fix in field pre/post process sections.
- Bug fix in reject section.

* Version 2.4-3 -- 1 November 2005

- Add VERSION labels for all modules.

* Version 2.4-2 -- 1 November 2005

- New root namespace: ETL::Pequel. 
- Initial Perl module program access interface via ETL::Pequel.pm. 
	-- Please view example/aggregates_1.pl for sample usage.
- New section types: 'display message on input/output', 'display message on input/output abort'.
	-- display a message to stderr if record matches condition(s), with optional processing abort.
- Support fixed length and packed input/output format via Perl pack/unpack formats 
	-- new options: 'unpack_input', 'input_pack_fmt', 'pack_output', 'output_pack_fmt'. 
- New 'copy input record', 'divert input record' section types replace 'copy/divert record'.
	-- distributed data processing.
- New 'copy output record', 'divert output record' section types.
- New macros: &pack(), &unpack().
- New macro: &input_record_count().
- Added gzcat_cmd, gzcat_args, cat_cmd, cat_args options.
- Added sort_cmd, sort_args, cpp_cmd, cpp_args options.
- New option 'input_record_limit' -- use to specify number of records to process from input.
- Additional example scripts demonstrate new copy/divert input/output record sections.
- Additional example scripts demonstrate new pack_output/unpack_input feature.
- Additional example scripts demonstrate new macros.
- Additional example scripts demonstrate new message section types.
- Fixed test failures.

* Version 2.3-4 -- 6 October 2005

- New 'copy record' section. Copy input record to second pequel script if condition met.
- New 'divert record' section. Divert input record to second pequel script if condition met.
- Additional example scripts demonstrate new copy/divert record sections.

* Version 2.3-3 -- 4 October 2005

- New section type 'sort output' -- sort the output from script.
- Now can use 'sort by' in chained pequel script.

* Version 2.3-2 -- 30 September 2005

- New option 'prefix' -- use to set directory path prefix for any file/script names specified in input_file,
	'load table', 'load table pequel', etc.
- Added Pequel script chaining via input_file -- receive input data from output of other Pequel script.
- Added 'load table pequel' section -- load a table via external Pequel script (see examples).
- Fixed code generation when using 'use_inline' with input_file specified.
- Fixed code generated for external tables with single data column.
- Added example scripts: pequel_tables.pql, sales_ttl_by_loc.pql.
- Removed Pequel::Base and 'root' references -- replaced by Pequel::Param.
- Extensive code cleanup and bug fixes.

* Version 2.2-9 -- 14 September 2005

- Added Vim syntax file.
- New tests.
- Additional example scripts.
- Revamped the table compiler -- now handles nested table and complex statements.
- Revamped the macro compiler -- now handles nested macros and complex statements.
- PEQUEL_TABLE_PATH env for runtime external tables path location.
- Must escape literal quote char as argument in Option with '\'. So: input_delimiter_extra(\"[)
- Allow quoted (single or double) Option argument.
- Fixed problem when group-by field references input derived field.
- Fixed code generation for arr_avg macro.
- All array macros now will parse any param as an array field.
- Fixed arr_size macro.
- Updated User Guide -- added BUGS list.
- Fixed various problems with examples scripts. Cleaned up examples dir.


* Version 2.2-8 -- 1 September 2005

- Updated User Guide.
- Updated User Guide -- added 'Inline Options' Chapter.
- Updated User Guide -- added Apache CLF example script.
- Added Apache Log (Common Log Format (CLF)) file parsing (quoted fields, square bracketed timestamp field).
- Added new option: input_delimiter_extra -- use for setting extra field delimiters, such as quotes and
	square brackets; eg: for Apache Log (Common Log Format (CLF)) file parsing specify input_delimiter_extra("[).
- Removed default settings for Inline => CCFLAGS and Inline => OPTIMIZE.
- Added new option: inline_ccflags -- use for setting CCFLAGS for Inline::C.
- Added new option: inline_optimize -- use for setting OPTIMIZE for Inline::C.
- Added new option: inline_cc -- use for settind CC for Inline::C.
- Fixed Inline Config=>NAME -- substitute '/' in path name by '::'.
- Added examples/apachelog.pql.


* Version 2.2-7 -- 25 August 2005

- Fixed User Guide incorrect reference to 'pql2' -- should be 'pequel'!
- Added examples directory containing sample pequel scripts together with their generated code.

lib/ETL/Pequel/Main.pm
#!/usr/bin/perl
# ----------------------------------------------------------------------------------------------------
#  Name		: ETL::Pequel::Main.pm
#  Created	: 14 January 2005
#  Author	: Mario Gaffiero (gaffie)
#
# Copyright 1999-2006 Mario Gaffiero.
# 
# This file is part of Pequel(TM).
# 
# Pequel is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# Pequel is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with Pequel; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
# ----------------------------------------------------------------------------------------------------
# Modification History
# When          Version     Who     What
# 02/03/2006	2.4-6b		gaffie	Fixed tests.
# 15/02/2006	2.4-6		gaffie	Fixed module name in ETL::Pequel::Pod2Pdf, scripts/pequelpod2pdf.
# 26/01/2006	2.4-6		gaffie	vim-syntax: fix description section.
# 01/12/2005	2.4-6		gaffie	fixed date.addUserType -- required PARAM.
# 30/11/2005	2.4-6		gaffie	exec_min_lines -- use exec() if code > lines.
# 16/11/2005	2.4-5		gaffie	Bug fix -- input record line counter.
# 15/11/2005	2.4-5		gaffie	Bug fix in sort-output section codeOpen() extra pipe.
# 11/11/2005	2.4-5		gaffie	new option show_synonyms -- 
# 04/11/2005	2.4-5		gaffie	Bug fix numeric/decimal type checking for nulls/nonulls options.
# 04/11/2005	2.4-5		gaffie	Bug fix numeric/decimal type comparison in dedup-on.
# 03/11/2005	2.4-4		gaffie	Bug fix field-process sections.
# 03/11/2005	2.4-4		gaffie	Bug fix in reject-section.
# 01/11/2005	2.4-2		gaffie	Fixed test failures -- caused by CPP parsing '#' comment.
# 26/10/2005	2.4-1		gaffie	ETL::Pequel.pm
# 26/10/2005	2.3-6		gaffie	display message section types.
# 26/10/2005	2.3-6		gaffie	&input_record_count() macros.
# 26/10/2005	2.3-6		gaffie	&arr_pack(), &arr_unpack() macros.
# 26/10/2005	2.3-6		gaffie	&pack(), &unpack() macros.
# 24/09/2005	2.3-6		gaffie	unpack_input/pack_output implementation.
# 19/10/2005	2.3-5		gaffie	Added gzcat_cmd, gzcat_args, cat_cmd, cat_args options.
# 19/10/2005	2.3-5		gaffie	Added sort_cmd, sort_args, cpp_cmd, cpp_args options.
# 13/10/2005	2.3-5		gaffie	New 'copy input record', 'divert input record' sections replace 'copy/divert record'.
# 13/10/2005	2.3-5		gaffie	New 'copy output record', 'divert output record' sections.
# 05/10/2005	2.3-4		gaffie	New 'copy record' section.
# 05/10/2005	2.3-4		gaffie	New 'divert record' section.
# 03/10/2005	2.3-3		gaffie	Allow sort-by when input-file option is pequel script.
# 03/10/2005	2.3-3		gaffie	New section type 'sort output'.
# 30/09/2005	2.3-2		gaffie	New option 'prefix' for prefixing input_file and sub pequel script names.
# 29/09/2005	2.3-2		gaffie	cmdType attribute to distinguish between cmdline and script only options.
# 27/09/2005	2.3-2		gaffie	Removed Pequel::Base usage and all refs to Pequel::root.
# 22/09/2005	2.3-2		gaffie	Parse.pm:Removed Pequel::Base usage and all refs to Pequel::root.
# 22/09/2005	2.3-2		gaffie	Begin removal of 'root'
# 21/09/2005	2.3-2		gaffie	Added -option option to view option values for script.
# 20/09/2005	2.3-2		gaffie	Added pequel script chaining functionality.
# 19/09/2005	2.3-2		gaffie	Added viewraw option for use by Pequel embedded tables.
# 15/09/2005	2.3-2		gaffie	Fixed code generation when using 'use_inline' with input_file specified.
# 14/09/2005	2.3-2		gaffie	Fixed code generated for external tables with single data column.
# 14/09/2005	2.3-2		gaffie	Added Pequel tables.
# 13/09/2005	2.2-9		gaffie	Docgen:Added About Pequel Chapter includes copyright notice;
# 13/09/2005	2.2-9		gaffie	Type/Table:PEQUEL_TABLE_PATH env for runtime external table path.
# 12/09/2005	2.2-9		gaffie	Parse/Parse:Revamped the table compiler
# 09/09/2005	2.2-9		gaffie	Type/Macro:Array field macros -- now assuming all args to be array field using &to_array().
# 09/09/2005	2.2-9		gaffie	Parse/Parse:Revamped the macro compiler -- now handles nested macros and complex statements.
# 05/09/2005	2.2-9		gaffie	Type/Section:Must escape litereal quote argument in Option with '\'.
# 05/09/2005	2.2-9		gaffie	Type/Section:Allow quoted (single/double) Option arguments
# 05/09/2005	2.2-9		gaffie	Engine/Engin:Prevent repeated calc of input field when they appear in group-by.
# 05/09/2005	2.2-9		gaffie	Engine/Engin:In codeBreak Fixed calc of derived group-by fields.
# 01/09/2005	2.2-9		gaffie	Bug fix in compile() for arr_avg macro.
# 01/09/2005	2.2-9		gaffie	All array macros now will parse any param as an array field.
# 01/09/2005	2.2-9		gaffie	Fixed arr_size macro.
# 01/09/2005	2.2-9		gaffie	Fixed example and test scripts containing old macro formats.
# 31/08/2005	2.2-8		gaffie	Added o_inline_lddlflags
# 31/08/2005	2.2-8		gaffie	Added o_inline_make
# 31/08/2005	2.2-8		gaffie	Added o_inline_libs, o_inline_inc.
# 31/08/2005	2.2-8		gaffie	Removed input_parse_input_qoutes option (replace by input_delimiter_extra).
# 31/08/2005	2.2-8		gaffie	Call cmdPrep method for all used options.
# 31/08/2005	2.2-8		gaffie	Type/Oprions:added cmdPrep method to Option::Element class.
# 31/08/2005	2.2-8		gaffie	Type/Oprions:added: input_delimiter_extra
# 30/08/2005	2.2-8		gaffie	Engine/Inline:added: inline_ccflags, inline_optimize
# 26/08/2005	2.2-8		gaffie	Engine/Inline:Fixed Config=>NAME -- subst '/' by '::'.
# 26/08/2005	2.2-8		gaffie	Engine/Inline:Added Apache Log file paring w/o_inline_parse_input_quotes via readsplit()
# 26/08/2005	2.2-8		gaffie	Engine/Inline:Added inline_cc option.
# 25/08/2005	2.2-7		gaffie	Code/Code.pm:updated sprintRaw so as to wrap code line > 110 chars.
# 25/08/2005	2.2-7		gaffie	Engine/Engine.pm:Added vim syntax perl setting in generated script.
# 25/08/2005	2.2-7		gaffie	Engine/Engine.pm:Removed 'use warnings' because eval in execute (might) complain.
# 25/08/2005	2.2-7		gaffie	Script/Script.pm:Bugfix with 'no cpp' misspelling.
# 25/08/2005	2.2-7		gaffie	Code/Code.pm:supress addCommentBegin()/End() unless --debug option specified.
# ----------------------------------------------------------------------------------------------------
# TODO:
# Qt GUI IDE for Pequel development/execution.
# PDF: pequel quick reference.
# PequelDTP: Web Data Transformation Portal.
# Combine common table loads across scripts.
# XML layer/interface: pequel-language -> XML -> pequel-engine.
# output_file(pequel:...) -- use this for output combiner, inherit attr so all divert/copy scripts will use.
# ETL::Pequel::Lang; ETL::Pequel::Lang::InputSection, ...; to contain parse(), anything pequel-language related.
# Sys::Syslog;
#
# DS1  DS2	 -- multi-datasources
# |    |
# PQL1 PQL2
#  \  /
#   \/
#  PQL3		-- combiner
#
# Need to use read/sysread for fixed length input when no eol character exists.
# Buffered output -- push output record to stack and output as batch -- enhance shared (locked) output.
# 	-- test output fd for lock; if locked continue processing next record.
# Option: input_record_length for use when input file is fixed length -- set $\ = record length;
# use pipe with syswrite to ensure that buffering does not deadlock the processes waiting for each other's message,
# use do-block instead of anonymous sub for macros that eval to anon-sub call.
# Replace 'use constant' in generated code with real-constants because use-constant replaces value with subroutine call() which degrades performance!
# apache: prefix for input/output_file spec.
# Socket pipe-fitting -- input_file(socket:...); output_file(socket:...); same with divert/copy section types.
# Use pack/unpack instead of substr in date manip code.
# input-section(file:, pequel:, oracle:, ...) for enabling multiple (successive) input stream
#	-- use input-merge-section(data-source, key-field) for multi simultaneous merge.
# dedup-input, dedup-output to replace dedup-on
# If packed input && sort-by then generate temp pequel srcipt and fit as input-file. The tmp script will do
# Combine Script.pm, Section->parse() functions, Parser.pm because they all relate to pequel-language parsing 
#	-- Parser/Pequel.pm; then future: Parser/XML.pm, etc
# Macro: &output_record_count()
# Macros: &input_fields_count(), &record_number(), 
# Pequel tables pack/unpack pipe interface.
# frequency analysis aggregate: &freq() -- ouput result in array field.
# Lazy pipe open; use fifo list for close() ordering -- only open copy/divert pipe if data matches condition.
# Replace pod2pdf with pdf package and rewrite pequeldoc stuff.
# merge/table cleanup:
# 	table(local: | file: | pequel: | oracle:connect_str | sqlite: | other:, static/dynamic, ...)
# 	merge(local: | file: | pequel: | oracle:connect_str | sqlite: | other:, ...)
#	input-merge section
# rename 'sort by' --> 'sort input'; output/input_file --> output/input_stream; 'input section'-->'input record'
# Perl module interface.
#   my $pequel = Pequel->new();
#	use constant GROUP_BY => 'group by';
#	$pequel->section(Pequel::GROUP_BY)->add(field => $fld, type => $type); -- will call parse() for Pequel::Type::Section::GroupBy;
#	or $pequel->group_by->add(...);
#	$pequel->section(Pequel::COPY_INPUT_RECORD, 'copy_ir_NSW.pql')->add(value => 'LOCATION =~ /NSW/') -- create derived section type;
#	or $pequel->copy_input_record('copy_ir_NSW.pql')->add(...);
#	or $pequel->copy_input_record($sub_pequel)->...;
#	$pequel->check();
#	$pequel->run();
# sort-by/sort-output with packing.
# pack/unpack interface between pequel scripts.
#		print map(pack("A3/Z*", $_), @VALS), "\n";
#		print join('|', unpack('A3/Z*' x 9, $_)), "\n";
# output_file(pequel:...)
# Fix divert output record when diverting to file.
# 'copy input record', 'copy output record', 'divert input record', ..., output_file(pequel:...);
# combiner: use named pipe
# 'distribute(thread_num)' -- split input based on condition and pipe to n processes.
#	how to avoid double splitting input record?
# 	implement as divert; remove input_file(), output_file(),
# &input_record() special macro acts on raw input line (ie before split) -- use in filter, etc sections.
# &num_fields() -- return number of fields in input record;
# input_format option: input_format(pequel:script.pql); input_format(delim:file.txt); 
#	apachelog:, excell:, cvs:, fixed:, ...
# 'reject record(file:outfile.dat)' section. Divert input record if matches condition.
# 'reject record(pequel:outfile.pql)' -- reject and divert are identical.
# Chain Pequel scripts -- 
#	input section not required -- get fields names from output_format of input_file script.
# Bug:Input derived fields referenced in output calc or where clause not being calculated.
# &to_array(..., delim);
# ----------------------------------------------------------------------------------------------------
require 5.005_62;
use strict;
use attributes qw(get reftype);
use warnings;
use vars qw($VERSION $BUILD);
$VERSION = "2.4-6b";
$BUILD = 'Thursday March 2 21:41:45 GMT 2006';
# ----------------------------------------------------------------------------------------------------
{
	package ETL::Pequel::Main;
	use UNIVERSAL qw( isa can );

	our $this = __PACKAGE__;
	sub BEGIN
	{
		our @attr =
		qw(
			PARAM
		);
		eval ("sub attr { my \$self = shift; return (qw(@{[ join(' ', @attr) ]})); } ");
		foreach (@attr)
		{
			eval
			("
				sub @{[ __PACKAGE__ ]}::$_ : method
				{
					my \$self = shift;
					\$self->{\$this}->{@{[ uc($_) ]}} = shift if (\@_);
					return \$self->{\$this}->{@{[ uc($_) ]}};
				}
			");
		}
	}

	sub new : method
	{
		my $proto = shift;
		my $class = ref($proto) || $proto;
		my $self = {};
		bless($self, $class);

		use ETL::Pequel::Param;
		$self->PARAM(ETL::Pequel::Param->new());
		$self->initRc();
	
		if (my $scriptname = shift) 
		{
			$self->PARAM->properties('script_name', $scriptname); 
		
			if (my $parent_PARAM = shift)
			{
				$self->PARAM->root($parent_PARAM->root());
				foreach my $o (grep($_->inherit, $self->PARAM->options->toArray()))
				{
					$self->PARAM->properties($o->name(), $parent_PARAM->properties($o->name()));
				}
				$self->PARAM->depth($parent_PARAM->depth() +1);
			}
			$self->prepare();
		}
		else
		{
			$self->PARAM->root($self);
			$self->parseCmdLine();
			$self->prepare();
			$self->generate();
		}
		return $self;
	}

	sub prepare : method
	{
		my $self = shift;

		$self->PARAM->SCRIPT->prepare();
		$self->PARAM->options->cmd_prep();
		$self->PARAM->options->cmd_exec();
	}

	sub generate : method
	{
		my $self = shift;

		$self->PARAM->ENGINE->generate();
		$self->PARAM->ENGINE->prepare();
		$self->PARAM->options->cmd_postgen();
	}

	sub execute : method
	{
		my $self = shift;
		if ($self->PARAM->properties('script_name') eq '')
		{
			$self->PARAM->SCRIPT->usage();
			return;
		}
		$self->PARAM->error->msgStderr("");
		if (!$self->PARAM->ENGINE->size())
		{
			$self->prepare();
			$self->generate();
#??			$self->execute();
		}
		$self->PARAM->ENGINE->execute();
	}

	sub check : method
	{
		my $self = shift;
		if ($self->PARAM->properties('script_name') eq '')
		{
			$self->PARAM->SCRIPT->usage();
			return;
		}
		if (!$self->PARAM->ENGINE->size())
		{
			$self->prepare();
			$self->execute();
		}
		return $self->PARAM->ENGINE->check();
	}

#>	Move to ETL::Pequel::Parser::CmdLine
	sub parseCmdLine : method 
	{
		# Cmdline option settings override:
		#	Script option settings, which override:
		#		Pequelrc option settings,

		my $self = shift;
		use Getopt::Long;

		my $sub = sub 
		{ 
			my $name = shift;
			my $arg = shift || undef;
			my $o;

			$self->PARAM->error->fatalError("Invalid option '$name'")
				if 
				(
					!($o = $self->PARAM->options->exists($name))
					&& !($o = $self->PARAM->options->getAlias($name))
				);

			$self->PARAM->sections->exists('options')->items->add(ETL::Pequel::Field::Element->new
			(
				name => $o->name(), 
				value => defined($arg) ? $arg : $o->cmdFormat() eq '!' ? 1 : $o->value(),
				comment => '_CMDLINE',
				ref => $o,
				PARAM => $self->PARAM
			));
		};

		my %Options;
		foreach (grep($_->cmdFormat, $self->PARAM->options->toArray()))
#>		foreach (grep($_->can("cmdFormat"), $self->root->t_option->toArray))
		{
			$Options{$_->name . $_->cmdFormat} = \&{$sub};
			next unless ($_->cmdAlias->size);
			foreach my $alias ($_->cmdAlias->toArray())
			{
				$Options{$alias->name . $_->cmdFormat} = \&{$sub};
			}
		}

		$self->PARAM->error->fatalError("Use --help option for info (@{[ join(',', @ARGV) ]}).")
			if (!Getopt::Long::GetOptions(%Options));

		if (!$self->PARAM->sections->exists('options')->items->exists('script_name') && grep(/.pql$/, @ARGV))
		{
			$self->PARAM->sections->exists('options')->items->add(ETL::Pequel::Field::Element->new
			(
				name => 'script_name',
				value => grep(/.pql$/, @ARGV),
				comment => '_CMDLINE',
				ref => $self->PARAM->options->exists('script_name'),
				PARAM => $self->PARAM
			));
		}
	}

	sub initRc : method
	{
		my $self = shift;
		# Open the pequel init file ~/.pequelrc and set any options
		if (-e "$ENV{HOME}/.pequelrc")
		{
			$self->PARAM->error->msgStderr("reading ~/.pequelrc...") if ($self->PARAM->properties('debug'));
			open(PequelRC, "$ENV{HOME}/.pequelrc");
# put <> into slurp mode
#undef $/;
## read configuration file supplied on command line into string
#my $configuration = <>;
#my %config;
## read all configuration options from config string
#while ($configuration =~ /^\s*(\w+)\s* = \s*(.+?)\s*$/mg) {
#  $config{$1} = $2;
#}
#print "Got: $_ => '$config{$_}'\n" foreach (sort keys %config);
			while (<PequelRC>)
			{
				chomp;
				last if (/__END__/);
				s/#.*//;		# remove comments
				s/\/\/.*//g;	# remove c style comment lines if not cpp'd
				s/^\s*//;
				s/\s*$//;
				s/\s*,$//;
				next if ($_ eq '');

				if (/^&|^macro/)	# its a macro
				{
					s/^&\s*|^macro\s*//;
					my $name = substr($_, 0, index($_, '=')-1);
					my $exp = substr($_, index($_, '=')+1, length($_)-index($_, '='));
					$self->PARAM->macros->add
					(
						ETL::Pequel::Type::UserMacro->new
						(
							name => $name, 
							eval => $exp,
							PARAM => $self->PARAM
						)
					);
				}
				elsif (/^aggregate/)	# its an aggregate
				{
				}
				elsif (/^datetype/)	# its a datetype
				{
				}
				else # assume its an option
				{
					my ($name, $value) = split(/\s*=\s*/);
					my $o;
					if 
					(
						($o = $self->PARAM->options->exists($name))
						|| ($o = $self->PARAM->options->getAlias($name))
					)
					{
						$self->PARAM->error->msgStderr("->$name=$value;") if ($self->PARAM->properties('debug'));
						$self->PARAM->sections->exists('options')->items->add(ETL::Pequel::Field::Element->new
						(
							name => $o->name, 
							value => $value,
							comment => '_PequelRC',
							ref => $o,
							PARAM => $self->PARAM
						));
					}
				}
			}	
		}
	}
}
1;
# ----------------------------------------------------------------------------------------------------
